sip-0xx-improve-sip10/sip-0xx-composable-fungible-token-standard.md

```diff
@@ -0,0 +1,151 @@
+
+```clarity
+(define-trait sip-0xx-trait
+  (
+    ;; Transfer from the caller to a new principal
+    ;; first principal, sender, must be always equal to contract-caller
+    (transfer (uint principal principal (optional (buff 34))) (response bool uint))
+
+    ;; the human readable name of the token
+    (get-name () (response (string-ascii 32) uint))
+
+    ;; the ticker symbol, or empty if none
+    (get-symbol () (response (string-ascii 32) uint))
+
+    ;; the number of decimals used, e.g. 6 would mean 1_000_000 represents 1 token
+    (get-decimals () (response uint uint))
+
+    ;; the balance of the passed principal
+    (get-balance (principal) (response uint uint))
+
+    ;; the current total supply (which does not need to be a constant)
+    (get-total-supply () (response uint uint))
+
+    ;; an optional URI that represents metadata of this token
+    (get-token-uri () (response (optional (string-utf8 256)) uint))
+
+    ;; Transfer from one principal to another using an allowance
+    (transfer-from (uint principal principal uint) (response bool uint))
+
+    ;; Approve an incremental allowance for a specific principal to spend tokens
+    (approve (principal uint) (response bool uint))
+
+    ;; Revoke an allowance, goes to 0, for a specific principal to spend tokens
+    (revoke (principal) (response bool uint))
+
+    ;; Check the remaining allowance of tokens for a spender
+    (allowance (principal principal) (response uint uint))
+  )
+)
+```
+
+This extended trait, `sip-0xx-trait`, includes the functions from the original `sip-010-trait` and adds the new functions introduced in SIP-0XX: `transfer-from`, `approve`, `revoke`, and `allowance`. Developers can use this trait as a reference when implementing composable fungible tokens with allowances on the Stacks blockchain.
+
+## Rationale
+
+The extension of the SIP-010 trait with allowances and the ability to transfer tokens using allowances addresses the limitations of the previous standard. By introducing allowances, users can grant explicit permission for third parties to spend tokens on their behalf, improving the security and composability of DeFi contracts. The inclusion of additional functions from SIP-010 ensures compatibility with existing standards.
+
+### Limiting Phishing 
+
+With this new approach that has only a check for `sender == contract-caller`, and in case of successful phishing attemp, the malicious Dapp has to ask for allowance for the specific token an drain that token, so the they have to request 2 transactions and can only drain 1 token. The previous standard, and the de-facto check of `sender == tx-sender`, and the phishing attemp is successful, with a single transaction they can drain all of our standard tokens (several contracts) that only check the `tx-sender`.
+
+### DeFi Composability Pattern
+
+The most common DeFi pattern, that is supported by this new standard is:
```

Comment:
I think it can be done with SIP10 with `transfer` only if tokens check `tx-sender`, if lets say DeFiA uses DeFiB and DeFiA transfer your tokens from your wallet to DeFiB. Then you give total control to the DeFiA and also to DeFiB called by DeFiA. Not sure your tokens can be retrieved properly because your tokens are jumping from your wallet to DeFiB.

Reply:
So, why do we need allowances? Is it only about the issue of tx with allow mode?

Reply:
if you have DeFiA is a YieldAggregator and DeFiB is a YieldService, then the following 3 current approaches does not work I think:

1)

a) User calls `defi-service` from DeFiA.
b) DeFiA does `transfer(User,DeFiA)`, good tx-sender.
c) In the same transaction DeFiA cannot do `transfer(DeFiA,DefiB)` because the tx-sender is still User.

2)

a) User calls `defi-service` from DeFiA.
b) DeFiA does `transfer(User,DeFiB)`, good tx-sender directly from User.
c) Incorrect, DeFiB does not supports receiving random token from third parties without calling `another-defi-service` from DeFiB.

3)

a) User calls `defi-service` from DeFiA.
b) DeFiA does `transfer(User,DeFiA)`, good tx-sender.
c) DeFiA calls `another-defi-service` from DeFiB.
d) Incorrect, tx-sender is still User, if DeFiB calls `transfer(DeFiA,DeFiB)` it has no permission.

Reply:
In 1. c) DeFiA can use `as-contract` to switch context, but needs to be careful. Furthermore, the token can allow transfers where sender is contract-caller OR tx-sender. Then no context switch is necessary.

In 3. c) Similarily, DeFiA can use `as-contract` to switch context, but needs to be careful.

In 3. DeFiA could not do b) (transfer tokens from User to DeFiA), but let DeFiB do it in d).

Reply:
okey, so the solution for you is to change `tx-sender` to local contract by using `as-contract`.

Reply:
@friedger https://forum.stacks.org/t/escrow-contract/10642
![image](https://github.com/stacksgov/sips/assets/30682875/79435163-c32e-49ed-9763-fad986c90bbf)